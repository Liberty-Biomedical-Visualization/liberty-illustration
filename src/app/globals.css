@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
  --accent-50: 224 247 252;
  --accent-100: 178 224 240;
  --accent-200: 128 200 232;
  --accent-300: 77 176 215;
  --accent-400: 26 153 206;
  --accent-500: 3 152 206;
  --accent-600: 2 128 165;
  --accent-700: 2 109 136;
  --accent-800: 2 90 107;
  --accent-900: 2 61 77;
  --accent-950: 1 44 58;
  --background: 255 255 255;
  --foreground: 23 23 23;
}

body {
  color: var(--foreground);
  background: var(--background);
  font-family: Arial, Helvetica, sans-serif;
}

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}
