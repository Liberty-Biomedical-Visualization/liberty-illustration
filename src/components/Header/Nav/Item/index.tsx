"use client";

import Link from "next/link";
import { usePathname } from "next/navigation";

import resolveClassNames from "@/lib/resolveClassNames";

export default function NavItem(props: Readonly<NavItemProps>) {
  const { children, href } = props;
  const pathname = usePathname();

  const color =
    pathname === href
      ? "active:text-accent-200 hover:text-accent-400 text-accent-600"
      : "active:text-accent-400 hover:text-accent-600 text-accent-800";

  const linkClassName = resolveClassNames(
    "text-lg transition uppercase",
    color,
    props.className,
  );

  return (
    <Link className={linkClassName} href={href}>
      {children}
    </Link>
  );
}

export interface NavItemProps {
  children: string;
  className?: string;
  href: string;
}
