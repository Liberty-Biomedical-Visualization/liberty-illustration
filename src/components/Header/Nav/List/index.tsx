import resolveClassNames from "@/lib/resolveClassNames";

import ListItem, { NavListItemProps as ItemProps } from "./Item";

export default function NavList(props: Readonly<NavListProps>) {
  const { onItemClick, pages } = props;

  const className = resolveClassNames(baseClassNames, props.className);

  const items = pages.map((page, index) => {
    const isLastPage = index === pages.length - 1;
    return transformToItem(page, isLastPage, onItemClick);
  });

  return <ul className={className}>{items}</ul>;
}

export interface NavListProps {
  className?: string;
  onItemClick: ItemProps["onClick"];
  pages: readonly Readonly<PageNavData>[];
}

const baseClassNames = "md:flex";

function transformToItem(
  page: Readonly<PageNavData>,
  isLastPage: boolean,
  onClick: NavListProps["onItemClick"],
) {
  const { href, name } = page;

  const className = isLastPage
    ? ""
    : "md:after:content-pipe md:after:mx-2 md:after:text-foreground";

  return (
    <ListItem className={className} href={href} key={href} onClick={onClick}>
      {name}
    </ListItem>
  );
}

interface PageNavData {
  href: ItemProps["href"];
  name: ItemProps["children"];
}
