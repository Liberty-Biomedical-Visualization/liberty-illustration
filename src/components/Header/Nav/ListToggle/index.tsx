import { faBars, faTimes } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import AccessiblyHiddenText from "@/components/AccessiblyHiddenText";
import resolveClassNames from "@/lib/resolveClassNames";

export default function NavListToggle(props: Readonly<NavListToggleProps>) {
  const { id, listIsShown, navListId, onClick } = props;
  const className = resolveClassNames(baseClassName, props.className);

  return (
    <button
      aria-controls={navListId}
      aria-expanded={listIsShown}
      aria-haspopup
      className={className}
      id={id}
      onClick={onClick}
    >
      <FontAwesomeIcon
        fixedWidth
        icon={listIsShown ? faTimes : faBars}
        size="2x"
      />
      <AccessiblyHiddenText>Navigation</AccessiblyHiddenText>
    </button>
  );
}

export interface NavListToggleProps {
  className?: string;
  id: string;
  listIsShown: boolean;
  navListId: string;
  onClick: () => void;
}

const baseClassName =
  "cursor-pointer fixed right-4 md:hidden py-0.5 top-4 z-20";
