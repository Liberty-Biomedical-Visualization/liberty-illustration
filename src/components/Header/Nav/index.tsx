"use client";

import FocusTrap from "focus-trap-react";
import { usePathname } from "next/navigation";
import { useCallback, useEffect, useRef, useState } from "react";

import resolveClassNames from "@/lib/resolveClassNames";

import NavList, { type NavListProps } from "./List";
import NavListToggle from "./ListToggle";

export default function Nav(props: Readonly<NavProps>) {
  const { disableTabbableDisplayCheck, pages } = props;
  const [navIsShown, setNavIsShown] = useState(false);
  const pathname = usePathname();
  const lastPathname = useRef(pathname);
  const bodyPaddingRight = useRef(0);

  const hideNav = useCallback(() => {
    enableScrolling(bodyPaddingRight.current);
    setNavIsShown(false);
  }, []);

  const hideNavOnNavigation = useCallback(() => {
    if (navIsShown && lastPathname.current !== pathname) {
      hideNav();
    }
  }, [hideNav, lastPathname, navIsShown, pathname]);

  useEffect(hideNavOnNavigation, [hideNavOnNavigation]);

  const updateLastPathname = () => {
    lastPathname.current = pathname;
  };

  useEffect(updateLastPathname, [pathname]);

  useEffect(
    () => hideNavWhenShownAboveMediumBreakpoint(hideNav, navIsShown),
    [hideNav, navIsShown],
  );

  const showNav = () => {
    bodyPaddingRight.current = parseFloat(
      getComputedStyle(document.body).paddingRight,
    );
    disableScrolling(bodyPaddingRight.current);
    setNavIsShown(true);
  };

  const toggleList = () => {
    if (navIsShown) {
      hideNav();
    } else {
      showNav();
    }
  };

  const handleEscape = () => {
    hideNav();
    return true;
  };

  const hideNavOnSamePage = (href: string) => {
    if (href === pathname) {
      hideNav();
    }
  };

  const className = resolveClassNames(
    baseClassNames,
    navIsShown ? "visible" : "hidden",
    props.className,
  );

  return (
    <FocusTrap
      active={navIsShown}
      focusTrapOptions={{
        escapeDeactivates: handleEscape,
        tabbableOptions: {
          // Setting displayCheck to "none" allows FocusTrap to be used with
          // jsdom for testing.
          displayCheck: disableTabbableDisplayCheck ? "none" : "full",
        },
      }}
    >
      <div>
        <NavListToggle
          id={navListToggleId}
          listIsShown={navIsShown}
          navListId={navId}
          onClick={toggleList}
        />
        <nav aria-labelledby={navListToggleId} className={className} id={navId}>
          <NavList onItemClick={hideNavOnSamePage} pages={pages} />
        </nav>
      </div>
    </FocusTrap>
  );
}

export interface NavProps {
  className?: string;
  disableTabbableDisplayCheck?: boolean | undefined;
  pages: NavListProps["pages"];
}

const baseClassNames =
  "bg-white fixed h-full left-0 md:flex md:h-auto md:static md:visible top-0 w-full z-10";

const navId = "nav";
const navListToggleId = "nav-list-toggle";

function disableScrolling(bodyPaddingRight: number) {
  const scrollbarWidth =
    window.innerWidth - document.documentElement.clientWidth;
  document.body.style.overflow = "hidden";
  document.body.style.paddingRight = bodyPaddingRight + scrollbarWidth + "px";
}

function enableScrolling(bodyPaddingRight: number) {
  document.body.style.overflow = "auto";
  document.body.style.paddingRight = bodyPaddingRight + "px";
}

function hideNavWhenShownAboveMediumBreakpoint(
  hideList: () => void,
  listIsShown: boolean,
) {
  // Guard against running an environment without a global window such as tests.
  if (
    typeof window === "undefined" ||
    typeof window.matchMedia !== "function"
  ) {
    return;
  }

  const mediaQuery = window.matchMedia("(max-width: 768px)");

  const handleMediaQueryChange = () => {
    if (listIsShown && !mediaQuery.matches) {
      hideList();
    }
  };

  mediaQuery.addEventListener("change", handleMediaQueryChange);

  const stopHidingListWhenShownAboveMediumBreakpoint = () =>
    mediaQuery.removeEventListener("change", handleMediaQueryChange);

  return stopHidingListWhenShownAboveMediumBreakpoint;
}
