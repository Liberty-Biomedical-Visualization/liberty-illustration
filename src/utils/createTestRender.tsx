import { render } from "@testing-library/react";
import type { JSX, JSXElementConstructor } from "react";

/**
 * Returns a function that allows rendering Component with partial props while
 * falling back to defaultProps for testing purposes.
 */
export default function createTestRender<C extends JSXElementConstructor<P>, P>(
  Component: JSXElementConstructor<P> & C,
  defaultProps: JSX.LibraryManagedAttributes<C, P>,
) {
  return function (props: Partial<P> = {}) {
    const resolvedProps: JSX.IntrinsicAttributes &
      JSX.LibraryManagedAttributes<C, P> = { ...defaultProps, ...props };
    return render(<Component {...resolvedProps} />);
  };
}
