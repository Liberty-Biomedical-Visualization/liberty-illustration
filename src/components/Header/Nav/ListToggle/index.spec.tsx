import { describe, expect, it, jest } from "@jest/globals";
import { screen } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";

import createTestRender from "@/lib/createTestRender";

import NavListToggle from "./index";

describe("NavListToggle", () => {
  it("should display a button", () => {
    renderNavListToggle();
    const result = screen.getByRole("button");
    expect(result).toBeVisible();
  });

  it('should display text content "Navigation"', () => {
    renderNavListToggle({ listIsShown: false });
    const result = screen.getByRole("button", { name: "Navigation" });
    expect(result).toBeVisible();
  });

  it("should not be expanded when listIsShown is false", () => {
    renderNavListToggle({ listIsShown: false });
    const result = screen.getByRole("button", { expanded: false });
    expect(result).toBeVisible();
  });

  it("should be expanded when listIsShown is true", () => {
    renderNavListToggle({ listIsShown: true });
    const result = screen.getByRole("button", { expanded: true });
    expect(result).toBeVisible();
  });

  it("should call onClick when clicked", async () => {
    const onClick = jest.fn();
    renderNavListToggle({ onClick });

    const user = userEvent.setup();
    const button = screen.getByRole("button");
    await user.click(button);

    expect(onClick).toHaveBeenCalled();
  });

  it("should pass className to the outermost element", () => {
    const className = "foo";
    const result = renderNavListToggle({ className });
    expect(result.container.firstChild).toHaveClass(className);
  });

  it("should pass id to the outermost element", () => {
    const id = "foo";
    const result = renderNavListToggle({ id });
    expect(result.container.firstChild).toHaveAttribute("id", id);
  });

  it("should match the snapshot", () => {
    const result = renderNavListToggle();
    expect(result.container).toMatchSnapshot();
  });
});

const renderNavListToggle = createTestRender(NavListToggle, {
  id: "nav-list-toggle",
  listIsShown: false,
  navListId: "nav-list",
  onClick: () => {},
});
