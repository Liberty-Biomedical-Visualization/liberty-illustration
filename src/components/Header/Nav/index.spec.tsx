import { describe, expect, it } from "@jest/globals";
import { screen } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";

import createTestRender from "@/lib/createTestRender";

import Nav from ".";

describe("Nav", () => {
  it("should display a navigation toggle button", () => {
    renderNav();
    const result = screen.getByRole("button", { name: "Navigation" });
    expect(result).toBeVisible();
  });

  it("should display navigation when toggle button is clicked", async () => {
    renderNav();

    const user = userEvent.setup();
    const listToggle = screen.getByRole("button", { name: "Navigation" });
    await user.click(listToggle);

    const result = screen.getByRole("navigation");
    expect(result).toBeVisible();
  });

  it("should not display navigation when toggle is not clicked", () => {
    renderNav();
    const list = screen.getByRole("navigation", { hidden: true });
    expect(list).not.toBeVisible();
  });

  it("should display a list of links from passed pages when list toggle is clicked", async () => {
    const pages = [
      { href: "/", name: "Home" },
      { href: "/portfolio", name: "Portfolio" },
    ];
    renderNav({ pages });

    const user = userEvent.setup();
    const listToggle = screen.getByRole("button", { name: "Navigation" });
    await user.click(listToggle);

    const list = screen.getByRole("list");
    expect(list).toBeVisible();

    const listItems = screen.getAllByRole("listitem");

    const homeLink = screen.getByRole("link", { name: "Home" });
    const containsHomeLink = listItems.some((listItem) =>
      listItem.contains(homeLink),
    );
    expect(containsHomeLink).toBe(true);

    const portfolioLink = screen.getByRole("link", { name: "Portfolio" });
    const containsPortfolioLink = listItems.some((listItem) =>
      listItem.contains(portfolioLink),
    );
    expect(containsPortfolioLink).toBe(true);
  });

  it("should pass className to the navigation element", () => {
    const className = "foo";
    renderNav({ className });
    const navigation = screen.getByRole("navigation", { hidden: true });
    expect(navigation).toHaveClass(className);
  });

  it("should match the snapshot", () => {
    const result = renderNav();
    expect(result.container).toMatchSnapshot();
  });
});

const renderNav = createTestRender(Nav, {
  disableTabbableDisplayCheck: true,
  pages: [],
});
