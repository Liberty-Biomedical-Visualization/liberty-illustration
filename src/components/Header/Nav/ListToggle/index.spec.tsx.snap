// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavListToggle should match the snapshot 1`] = `
<div>
  <button
    aria-controls="nav-list"
    aria-expanded="false"
    aria-haspopup="true"
    class="cursor-pointer fixed md:hidden py-0.5 right-4 top-4 z-20"
    id="nav-list-toggle"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-bars fa-fw fa-2x "
      data-icon="bars"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 96C0 78.3 14.3 64 32 64l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 128C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 288c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32L32 448c-17.7 0-32-14.3-32-32s14.3-32 32-32l384 0c17.7 0 32 14.3 32 32z"
        fill="currentColor"
      />
    </svg>
    <span
      class="absolute border-0 h-px m-0 overflow-hidden p-0 w-px whitespace-nowrap"
    >
      Navigation
    </span>
  </button>
</div>
`;
